### PR TITLE
Add functionality which close the main menu on click outside of it.

### PR DESCRIPTION
**Pull request by grzegorzpokorski**
Title self-explanatory.

To achieve that added `data-name` attribute to Hamburger and NavLinks components. Then added function in useEffect hook in Nav component which handle functionality mentioned in title.

I am not sure that i assign appropriate types to the arguments. Maybe i should do it in different way but i do not know how. For example i do not know that i assigned a good type to the 'target' argument in function but 'lint-staged' passes commit only when i assigned them 'any' type 😬

### CODE REVIEW

**Comment by grzegorzpokorski** [comment]
Hi @robert-orlinski 👋🏻  I'm happy that you replied to me with such a tons of useful informations!

I knew that my code is not good enough but i didn't aware that it so much 🤔. I'm grateful to you that you mentioned what i did wrong and in the same time you pointed how can i deliver a better (more React way) solution by fixing a bugs and looking at the problem from the different angle.

To be honest i was thinking that implementation of this functionality will be much easier. I realised that i need learn more about 'React' and built in it hooks.

For now i haven't appropriate skill and knowelage to implement solution in the right way. I will try to implement that in more reusable and 'react' way, but it will take some time - myself don't know how long, mabe you will be faster 😀.
I decided to revert done changes in this PR and move it to the draft until i will be able to deliver a better solution.

One more time i want to emphasise that i'm so glad that you had a will to devote some of your private time for review my code and give ma such a contentfull feedback. Maybe on different occasion i will delivery not only an idea (and additional work by explaining my bugs to me 😬) but also a complex solution 😊.

I'm going to earn more knowledge ✌🏼

File: components/Nav/Nav.tsx [hunk 1]
```diff
@@ -20,12 +21,34 @@ const Nav = () => {
...
 
+  useEffect(() => {
+    const handleClickOffTheMenu = ({ target }: any) => {
+      const menu = document.querySelector('[data-name=main-menu]');
+      const menuOpener = document.querySelector('[data-name=main-menu-opener]');
+
+      do {
+        if (target === menu || target === menuOpener) return;
+        target = target.parentNode;
+      } while (target != null);
```
[the remark this answers is not in the document]
Comment (grzegorzpokorski) [reply components/Nav/Nav.tsx]: I didn't know that js has this beautifull function for that purpose 😍

**Comment by robert-orlinski** [comment]
So good to hear such words from you @grzegorzpokorski! 🥰 

My pleasure! I am happy that you could get some new, useful knowledge because of my comments and I am happy for your attitude - it's always awesome to do some code review and see that the person who is the author of reviewed code, appreciates the feedback and will be better dev because of it 🥳

For now, I will leave this idea and the PR for you - looking at your comments and your feedback after the review, I think it can be nice exercise for you to implement this idea in (as you said) more React way.

If in the end, you will not want to implement it in the future, I would be glad for info from you that you won't do it 😌

Thank you for this discussion, one more time thank you for the PR, and wish you all the best and effective learning! 🌱

**Comment by grzegorzpokorski** [comment]
Hi @robert-orlinski ☺️

I was thinking a lot about implement that functionality by creating dedicated component as you mention it above but i think a good approach will be also custom hook e.g `useOnClickOutside`.

My implementaion of this hook allows to trigger function with custom actions on click / touch outside of passed reference element to this custom hook.

In my opinion that gives us more flexibility what we want to be done after click / touch outside becouse of possibility to pass custom actions. This approach also deliver reusable solution ✨

What do you think about this 🤔

PS. Also this time i had a problem what type should i assign to specyfic elments so i will be grateful if you check them 🧐

**Comment by grzegorzpokorski** [comment]
Sorry, now i realized that dedicated component will be much cleaner approach and deliver **easier to read code**. I going to rebuild latest changes:
- change custom hook to return a bool value depending on if was clicked away of ...,
- dedicated compoment - move action  handling action which trigger by click outside of that compoment to component `ClickAwayListener`.

**Comment by grzegorzpokorski** [comment]
I have just pushed code with solution which use dedicated component `ClickAwayListener` and it works as i wanted :)

Only one issue apeared - namely this custom component return a div which wraps own childrens what could be cause problems with page styles in some cases.

I'm going to consider how to replace that with something what doesn't cause changes in page html structure.

For now i think solution is done - not perfect becouse of issue which i mention above, but i think it is not bad 😉

Of course i'm open for sugestions what can i improve.

**Comment by grzegorzpokorski** [comment]
Hello @robert-orlinski !

Short update what i have done:
- finaly created `ClickAwayListener` component which allows to detect event when we click away of contained element and triggers function which is passed to it as a prop. The component uses a newly created custom hook `useOnClickAway`.
Similar to  [MUI's ClickAwayListener](https://mui.com/base/react-click-away-listener/) which you have mentioned in a comment above, this component works only when the wrapped element is able to hold ref. Doesn't add additional HTML tags until wraps only one child, otherwise wraps childrens with `div` tag.
- created custom hook named `useOnClickAway`, which allows to detect click away from element, which is passed as a reference in argument and trigger a custom function which also are passed as a argument to this function. Motivation to create this custom hook was willing to provide a functionality which can be used in cases  where 'ClickAwayListener' component has limited powers - for e.g. in case when we want to detect click away event on functional components which require to use forwardRef method to hold ref.

I think all suggestions were implemented. The component is reusable. Also, `useOnClickAway` hook is reusable and can be used in cases where `ClickAwayListener` component can't.

Maybe you have other thoughts what can be improved or something should be done in different ways?

**Comment by robert-orlinski** [comment]
Hello @grzegorzpokorski, thank you for all your work, code written, and your comments! 

I've read your comments and new code, and have 2 "broader" questions before I go into the details. After finishing this comment, I will do some quick code review of the implementation details, as you requested 🤗

1. I can't get why we need this custom hook here. Wouldn't just using `useEffect()` which is literally the content of custom hook, be perfect thing to just place in `ClickAwayListener` component? 
2. This code:
```ts
if (React.Children.count(children) == 1) {
    const copyWithRef = React.cloneElement(children as ReactElement, { ref: ref });
    return <React.Fragment>{copyWithRef}</React.Fragment>;
}
```
looks for me like unnecessary. No matter if we wrap 1 DOM node in `<div />` or we wrap more, this code:
```ts
return <div ref={ref as RefObject<HTMLDivElement>}>{children}</div>;
``` 
should be ok (maybe beside of interpolation which can be removed if you remove this `if(React.Children...` and change `ref`'s type to `HTMLDivElement | null` for instance).

Now, I jump to the code!

File: components/ClickAwayListener/ClickAwayListener.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  onClickAway: () => void;
+};
+
+const ClickAwayListener: FC<Props> = ({ children, onClickAway }) => {
+  const ref = useRef<HTMLElement>(null);
+  useOnClickAway(ref, () => onClickAway());
+
+  if (React.Children.count(children) == 1) {
+    const copyWithRef = React.cloneElement(children as ReactElement, { ref: ref });
+    return <React.Fragment>{copyWithRef}</React.Fragment>;
```
Comment (robert-orlinski) [inline components/ClickAwayListener/ClickAwayListener.tsx:14]: You can just use `<></>` as a shorthand of `<React.Fragment></React.Fragment>` 🙂

File: components/ClickAwayListener/ClickAwayListener.tsx [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  return <div ref={ref as RefObject<HTMLDivElement>}>{children}</div>;
```
Comment (robert-orlinski) [inline components/ClickAwayListener/ClickAwayListener.tsx:17]: In general, if we think we have to use type assertion in TS (tell that "this value here is for 100% good, please be silent mr. compiler" 😅), for 99% our code can be improved or even redesigned. 

In general, good goal while writing TS is to avoid interpolation 99% of the time (there are exceptions like [narrowing types while using DOM operation methods like `document.querySelector()`](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions=)).

File: components/Nav/Nav.tsx [hunk 4]
```diff
@@ -20,27 +21,33 @@ const Nav = () => {
...
     <NavBar {...{ isSticky }}>
       <HiddenMainPageTitle />
       <Brand />
-      <Container>
-        <NavLinks areVisible={isNavVisibleOnMobile}>
-          <NavMenu />
-          <SocialMedia />
-        </NavLinks>
-        <SeparatorHiddenOnMobile />
+      <ClickAwayListener onClickAway={handleClickAway}>
```
Comment (robert-orlinski) [inline components/Nav/Nav.tsx:32]: IMHO, writing just `onClickAway={() => isNavVisibleOnMobile && toggleNavVisibility()}` would be faster to read than separate function. What do you think?

File: hooks/useOnClickAway.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+const useOnClickAway = (ref: RefObject<Node>, action: () => void): void => {
+  useEffect(() => {
+    const handleAction = (event: MouseEvent | TouchEvent) => {
+      if (!ref.current || ref.current.contains(event.target as Node)) return;
+      action();
+    };
+
+    document.addEventListener('mousedown', handleAction);
+    document.addEventListener('touchstart', handleAction);
```
Comment (robert-orlinski) [inline hooks/useOnClickAway.ts:11]: Isn't it true that `mousedown` pretty much equals `touchstart` on mobile devices? If I am right, we don't have to handle `touchstart` - `mousedown` should be enough.

File: hooks/useOnClickAway.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { useEffect, RefObject } from 'react';
+
+const useOnClickAway = (ref: RefObject<Node>, action: () => void): void => {
+  useEffect(() => {
+    const handleAction = (event: MouseEvent | TouchEvent) => {
+      if (!ref.current || ref.current.contains(event.target as Node)) return;
```
Comment (robert-orlinski) [inline hooks/useOnClickAway.ts:6]: We can simply check if `event.target` exists instead of asserting its type `as Node`.

**Review by robert-orlinski — CHANGES_REQUESTED** [review]

**Comment by grzegorzpokorski** [comment]
Hi @robert-orlinski 🙂

I going to try to explain your two questions as one because I think they are connected to each other.

As you were reading my comments which i wrote through the process of creating this code you probably noticed that, I thought that wrapping elements in this way:

`return <div ref={ref as HTMLDivElement | null}>{children}</div>;`,

will be good enough. But by manualy testing that soulition i realized, becouse of it added aditional `<div>` tag, in some cases it break styles. After that I decided to pass a ref to the first child of `ClickEvenetListener` component to make functionality of click away working with not breaking styles (that is a reason why I checking if `ClickAwayListener` wraps only one child and depending on that i copy ref to the child or wrap all childrens with additional `div`).

In my opinion this approach is slighty better becouse we don't need thinging about `ClickAwayListener` component when we write a styles and it doesn't affect html structure of the page but only add expected functionality.

I separate a part of code to custom hook called `onClickAway` becouse i used `React.cloneElement` and pass a ref to copied element what can be Functional Component. Functional components cannot be given refs so we can't use `ClickAwayListener` component becouse it copy first child which must be able to hold a ref (the same limitation which has a `ClickAwayListener` component from [MUI](https://mui.com/base/api/click-away-listener/)). In described situation custom hook `onClickAway` can be used in combination with `React.forwardRef`. I just leaved that hook to don't repeat that part of code which is the same in all cases when we want add some functionality on click away of some element.

I don't know what, you think about that?

File: components/Nav/Nav.tsx (hunk 4 above)
Comment (grzegorzpokorski) [reply components/Nav/Nav.tsx:32]: It looks much cleaner :) i will replace it.

File: components/ClickAwayListener/ClickAwayListener.tsx (hunk 2 above)
Comment (grzegorzpokorski) [reply components/ClickAwayListener/ClickAwayListener.tsx:14]: Thats true. It stays in this form after i tried different approaches including passing some props to React.Fragment. I will change that ✌

File: hooks/useOnClickAway.ts (hunk 5 above)
Comment (grzegorzpokorski) [reply hooks/useOnClickAway.ts:11]: Maybe you are right. That additional eventListener is form over substance and isn't required in this case - just listen on mousedown event is enough.

**Comment by grzegorzpokorski** [comment]
I have made most requested changes but not all yet. To be honest I have no idea how to improve that code especialy types. Current shape of code is only one that I found and doesn't rise any errors in code editor.

Maybe you have a right that it should be rebuild but how to rewrite this code to handle cases listed below, minimalise interference in the html structure and doesn't break styles of page, and at the same time make it as reusable as it can be?

I consider following cases:
- wrapping component or any element which can hold a ref, by `ClickAwayListener` (without adding additional `div` element, just pass a ref to first child of `ClickAwayListener` component),
- wrapping more than one component by `ClickAwayListener` (add additional `div` to  wrap all `ClickAwayListener` childrens - it can break styles),
- reuse `useOnClickAway` hook in case when I want wrap a functional component which uses `React.ForwardRef` to hold ref (similar to first case).

Perhaps I should consider less cases and focus on specific scenario? It could make that solution more simpler, less confusing but it will has more limitations 🤔

**Comment by robert-orlinski** [comment]
Hello @grzegorzpokorski! 

TLTR: Your solution is good in my opinion. I would do it in a simpler way but if you don't want to spend more time on this topic, we can just push code made by you already 🙂 You can only let me know if you want me to try to improve this typing. 

---

Thank you so much for those 2 comments and additional changes! Now I feel that I fully understand you and understand solutions proposed by you!

In general, to add some context to all my comments' and make sure we understand each other, maybe I should mention something similar to the question you pose here (mention it 10 or something comments ago):

> Perhaps I should consider less cases and focus on specific scenario? It could make that solution more simpler, less confusing but it will has more limitations 🤔

I feel that your solution is too big for the problem you are solving here.

Because of it, we have all these places with type interpolation, we have additional hook, we have ifs in `ClickAwayListener` component and all other things which we are talking about here. 

There is this [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle which is accurate here I think. It's better to do clean, easy to read and change solution than the one that solves every problem that can occur in the future. Simply because we can't anticipate every problem that can occur in the future 😅

**In this case simplest solution could be to:**

- Put this additional `div` as an element returned from `ClickAwayListener`. Just don't care about styling if it's not the case - as I see it's true for now. It looks for me that if we will use this additional `div`, we don't have to worry also about `ClickAwayListener`'s children - there could be multiple elements as children, functional component, class component, etc. - doesn't matter because we are creating and using `ref` in the `ClickAwayListener` component itself. 
- Put code from additional hook, also in the `ClickAwayListener` component. I don't see scenario in which we will need this hook as separate one, to put it in other components than `ClickAwayListener`. Putting this hook in other components than `ClickAwayListener` would be even bad thing IMHO because it will get messy if we are going to do the same thing with hook and the component.

But to make it also clear - for me current solution is also good, maybe even better than the one I am talking about (because it takes into consideration every scenario it can come to mind).

**Because of it, it's up to you:**
- You can tell me that you don't want to polish this code so, so much or you just don't agree with me that simpler solution will be better; add additional comments in this PR and just push it. In this case, let me know and I will push it. 
- You can also be convinced that simpler solution will be better here and implement it by looking into points above ("simplest solution can be to:").

**Additionally, you can let me know if you want me to improve typing which we were talking about 🙂** I will be happy to do it! As you said, you don't have idea on how to improve them, so maybe I can try 🥳

**Comment by grzegorzpokorski** [comment]
Hi @robert-orlinski

I think the proposed solution by me is close to handle all cases which came to my mind which can occur when we want to use functionality to detect click away some element ... but you have a right, this solution might be a little bit over engineered and maybe will need a separate documentation to understand what it does and in what combination it should be used to achieve the intended target 😅

So you convinced me to make it simpler, especially for the reason that simpler solution which you describe will deliver a solution for considered problem and will be more intuitive to use and less spectrum of cases makes it easier to add correct types in typescript particularly because I started using typescript short time ago.

For now a simpler solution will be quite well fitted to current needs as you said between the lines. In the future, I will look for a solution how to eliminate that additional div tag from the DOM or add additional improvements 🤔

In the upcoming days I'm going to deliver simpler solution 🙂

**Comment by robert-orlinski** [comment]
Hello @grzegorzpokorski 🙂

Thank you for this comment, I agree with everything you have said. I can't wait for the updated solution then!

If case you will need anything from me, let me know please 🤗

**Comment by grzegorzpokorski** [comment]
I deliver a simpler solution about what we talked recently. Before you decide to merge this 'PR' to main branch I want to share some thoughts which I collect after study whole process of searching / coding solution to make sentence from the title of this PR real.

I didn't know that providing this reusable solution will be as complex 😯 I have learned so much searching for the best solution - also because of your patiently help 💪 At the end I realised that sometimes it's good to make things simpler and don't reinvent wheel again. 

And here I think it would be a good time to mention that I recently start thinking that the better, simpler and more flexible solution for any cases when we need to add functionality which will be triggered by clicking outside of some element is simple custom hook instead dedicated component 🤯

This is a reason why I started thinking this way: according to react documentation, [components](https://en.reactjs.org/docs/components-and-props.html) are used in describing what should appear on the screen, while [hooks](https://en.reactjs.org/docs/hooks-intro.html) are used as one of the source of logic in the application - both are used in different layers of application (logic / UI).

Just look, if we will use hook instead of dedicated component problem with additional (redundant (?)) HTML tags disappears because only thing what we need to do is simply add a ref to the container (e.g navigation container), and pass it with a function to hook which will handle passed fucntion when user click away from the container. That approach also delegates problem with many nodes or functional components, which need React.ForwardRef to hold ref, to programmer who create components.

But at least a dedicated ClickAwayListener component looks clean and easy to use like some handy tool 🛠

To sum up I think the current state of this PR is ready do deploy, but it has some limitation and I wonder if this state is caused because of wrong approach to solve problem 🤔

**Comment by grzegorzpokorski** [comment]
@robert-orlinski, I am so thankful for your time and patience, which you dedicated to teach me something. 😊 I am so happy because of a lot of all your comments, suggestions that help make me be better developer and motivated me to start learning typescript. One more time THANK YOU! 😁👋

**Comment by robert-orlinski** [comment]
@grzegorzpokorski, my pleasure 🥰 So good to hear that!
